l3kernel/l3pdf.dtx

```diff
@@ -314,52 +363,248 @@
 %
 % \subsection{Objects}
 %
+% \begin{variable}{\g_@@_backend_object_int}
+%   For returning object numbers.
+%    \begin{macrocode}
+\int_new:N \g_@@_backend_object_int
+%    \end{macrocode}
+% \end{variable}
+%
 % \begin{macro}{\pdf_object_new:n}
 % \begin{macro}
 %   {\pdf_object_write:nnn, \pdf_object_write:nne, \pdf_object_write:nnx}
 % \begin{macro}{\pdf_object_ref:n}
-% \begin{macro}
-%   {
-%     \pdf_object_unnamed_write:nn, \pdf_object_unnamed_write:ne,
-%     \pdf_object_unnamed_write:nx
-%   }
-% \begin{macro}{\pdf_object_ref_last:}
+% \begin{macro}{\__kernel_pdf_object_id:n}
+% \begin{macro}{ltx.pdf.object_id}
+% \begin{macro}{\@@_object_record:nn, \@@_object_record:nV}
+% \begin{macro}{\@@_object_retrieve:n}
 % \begin{macro}[pTF]{\pdf_object_if_exist:n}
 %   Simple to do: all objects create a constant |int| so it is not a
 %   backend-specific name.
 %    \begin{macrocode}
 \cs_new_protected:Npn \pdf_object_new:n #1
   {
-    \exp_args:Ne \@@_backend_object_new:n {#1}
+    \@@_backend_object_new:
+    \@@_object_record:nV {#1} \g_@@_backend_object_int
   }
 \cs_new_protected:Npn \pdf_object_write:nnn #1#2#3
   {
-    \exp_args:Ne \@@_backend_object_write:nnn {#1} {#2} {#3}
+    \exp_args:Ne \@@_backend_object_write:nnn
+      { \@@_object_read:n {#1} } {#2} {#3}
     \bool_gset_true:N \g_@@_init_bool
   }
 \cs_generate_variant:Nn \pdf_object_write:nnn { nne , nnx }
-\cs_new:Npn \pdf_object_ref:n #1 { \@@_backend_object_ref:n {#1} }
-\cs_new_protected:Npn \pdf_object_unnamed_write:nn #1#2
+\cs_new:Npn \pdf_object_ref:n #1
   {
-    \exp_args:Ne \@@_backend_object_now:nn {#1} {#2}
-    \bool_gset_true:N \g_@@_init_bool
+    \exp_args:Ne \@@_backend_object_ref:n
+      { \@@_object_retrieve:n {#1} }
   }
-\cs_generate_variant:Nn \pdf_object_unnamed_write:nn { ne , nx }
-\cs_new:Npn \pdf_object_ref_last: { \@@_backend_object_last: }
+\cs_new:Npn \__kernel_pdf_object_id:n #1
+  {
+    \exp_args:Ne \@@_backend_object_id:n
+      { \@@_object_retrieve:n {#1} }
+  }
+%</tex>
+%    \end{macrocode}
+%   Object mappings are tracked in Lua for \LuaTeX{} as this makes retrieving
+%   them much easier; as a result, there is a split in approaches. In Lua we
+%   store values in a table indexed by name.
+%    \begin{macrocode}
+%<*lua>
+
+local scan_int = token.scan_int
+local scan_string = token.scan_string
+
+local @@_objects = {}
+
+luacmd('@@_object_record:nn', function()
+  local name = scan_string()
+  local n = scan_int()
+  @@_objects[name] = n
+end, 'global')
+
+local function object_id(name)
+  return @@_objects[name] or 0
+end
+
+luacmd('@@_object_retrieve:n', function()
+  local name = scan_string()
+  return object_id(name)
```

Comment:
You probably want something like
```suggestion
  tex.cprint(12, tostring(object_id(name)))
```
here, otherwise the return value is ignored.

Reply:
Adjusted (I still don't really get all of the `print` variants Hans provides!)